lamindb/dev/db/_add.py

```diff
@@ -81,28 +81,93 @@ def add(  # type: ignore
             return results
         else:
             records = [model(**fields)]
-    for record in records:
-        if isinstance(record, DObject) and hasattr(record, "_local_filepath"):
-            upload_data_object(record, use_fsspec=use_fsspec)
+
     if session is None:  # assume global session
         session = settings.instance.session()
         settings.instance._cloud_sqlite_locker.lock()
         close = True
     else:
         close = False
+
+    # commit metadata to database
+    raised_errors = []
     for record in records:
+        prepare_filekey_metadata(record)
         session.add(record)
-    session.commit()
-    for record in records:
-        session.refresh(record)
+    try:
+        session.commit()
+    except Exception as e:
+        raised_errors.append(e)
```

Comment:
I'm not entirely sure why you'd want to continue with the execution of the function in this case - let's simply discuss this in our meeting later today!

If there is an error during `session.commit()` in line 98, I'd simply close the session and exit the function rather than going into the "refresh & upload section".

Reply:
Agreed - in fact, I initially implemented it as such!

Your previous comment on appending error messages made me think you did not want to exit the function after a first error.